### PR TITLE
Set application version properties for Maui projects

### LINKED ISSUE #828: Versioning issue with MAUI
Using `Nerdbank.GitVersioning` together with MAUI is not that easy with regards to setting `ApplicationVersion` and `ApplicationDisplayVersion` properties. The main problem is setting the properties after `Nerdbank.GitVersioning` has exceuted, but before generating the manifests for Android, iOS and Windows.

The following code takes care of all this, and also adheres to the versioning requirements for each platform. The targets `FixVersionFor*` will make sure `SetApplicationVersion` is executed at the right time in the MSBuild pipeline.

NOTE: This code only takes care of Android, iOS and Windows.

``` xml
    <Target Name="FixVersionForAndroid"
            Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'"
            BeforeTargets="_GetAndroidPackageName"
            DependsOnTargets="SetApplicationVersion" />

    <Target Name="FixVersionForIOS" 
            Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'"
            BeforeTargets="_CompileAppManifest"
            DependsOnTargets="SetApplicationVersion" />

    <Target Name="FixVersionForWindows" 
            Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'" 
            BeforeTargets="MauiGeneratePackageAppxManifest" 
            DependsOnTargets="SetApplicationVersion" />

    <Target Name="SetApplicationVersion" DependsOnTargets="GetBuildVersion">

        <PropertyGroup>
            
            <_MajorStr>$([System.String]::Format("{0}", $([System.Version]::Parse("$(BuildVersion)").Major)))</_MajorStr>
            <_MinorStr>$([System.String]::Format("{0:D2}", $([System.Version]::Parse("$(BuildVersion)").Minor)))</_MinorStr>
            <_BuildStr>$([System.String]::Format("{0:D3}", $([System.Version]::Parse("$(BuildVersion)").Build)))</_BuildStr>

            <!-- Android requirement: ApplicationVersion must be a positive integer (used as an internal version number) -->
            <!-- iOS requirement: ApplicationVersion must be a three part version number -->
            <!-- Windows requirement: ApplicationVersion must be blank when ApplicationDisplayVersion is a four part version number -->
            <ApplicationVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">$(_MajorStr)$(_MinorStr)$(_BuildStr)</ApplicationVersion>
            <ApplicationVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">$(BuildVersionSimple)</ApplicationVersion>
            <ApplicationVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'"></ApplicationVersion>

            <!-- iOS requirement: ApplicationDisplayVersion must be a three part version number -->
            <!-- Windows requirement: ApplicationDisplayVersion must be a four part version number -->
            <ApplicationDisplayVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">$(Version)</ApplicationDisplayVersion>
            <ApplicationDisplayVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">$(BuildVersionSimple)</ApplicationDisplayVersion>
            <ApplicationDisplayVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">$(BuildVersion)</ApplicationDisplayVersion>

        </PropertyGroup>

    </Target>
```

### PR DESCRIPTION
**Pull request by AArnott**
Closes #828

### CODE REVIEW

**Comment by AArnott** [comment]
@espenrl Can you confirm that this works for you? If you can't confirm from this PR directly, you should be able to consume from our CI feed once this PR is merged as per the [readme instructions](https://github.com/dotnet/Nerdbank.GitVersioning#how-do-i-consume-the-latest-changes-prior-to-their-release-on-nugetorg).

File: src/Nerdbank.GitVersioning.Tasks/build/Nerdbank.GitVersioning.targets [hunk 1]
```diff
@@ -219,6 +219,56 @@
...
+      <_NBGV_Major_Shifted>$([MSBuild]::Multiply($([System.Version]::Parse('$(BuildVersion)').Major), 16777216))</_NBGV_Major_Shifted>
+      <_NBGV_Minor_Shifted>$([MSBuild]::Multiply($([System.Version]::Parse('$(BuildVersion)').Minor), 65536))</_NBGV_Minor_Shifted>
+
+      <ApplicationVersion>$([MSBuild]::Add($([MSBuild]::Add($(_NBGV_Major_Shifted), $(_NBGV_Minor_Shifted))), $(BuildNumber)))</ApplicationVersion>
+      <ApplicationDisplayVersion>$(Version)</ApplicationDisplayVersion>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="NBGV_FixVersionForIOS"
+          Condition="'$(TargetPlatformIdentifier)' == 'ios'"
```
Comment (espenrl) [inline src/Nerdbank.GitVersioning.Tasks/build/Nerdbank.GitVersioning.targets:250]: $(TargetPlatformIdentifier)' == 'ios' Or '$(TargetPlatformIdentifier)' == 'maccatalyst'

I've just tested that IOS and MacOS are the same.

**Comment by espenrl** [comment]
It's great you're including MAUI support into the library. I guess one must take into account MacOS and Tizen also to cover the full range of MAUI. I've added a comment about MacOS. I will investigate Tizen.

**Comment by espenrl** [comment]
It seems like MAUI Tizen doesn't use `ApplicationVersion` / `ApplicationDisplayVersion` at all. It simply reads what's in the manifest file `Platforms/Tizen/tizen-manifest.xml`

So while the following code will alter the version in the output package filename (`TizenPackageVersion`), I'll have to guess the manifest is embedded with another version. Major issue. This is is stark contrast to Windows, iOS, MacOS and Android where the manifest file is updated according to `ApplicationVersion` / `ApplicationDisplayVersion`.

Let's skip Tizen support for now.

``` xml
<Target Name="NBGV_SetVersionfForMauiTizen"
          Condition="'$(TargetPlatformIdentifier)' == 'tizen'"
          BeforeTargets="_TizenResolveTpkPackageName"
          DependsOnTargets="GetBuildVersion">
    <PropertyGroup>
      <ApplicationVersion>$(BuildVersionSimple)</ApplicationVersion>
      <ApplicationDisplayVersion>$(BuildVersionSimple)</ApplicationDisplayVersion>
      <TizenPackageVersion>$(BuildVersionSimple)</TizenPackageVersion>
    </PropertyGroup>
</Target>
```

Please see _TizenGetTpkInfoFromManifest

https://github.com/Samsung/Tizen.NET/blob/4e822dbbe2f9d483badc608257a9e3ca9118010d/workload/src/Samsung.Tizen.Sdk/targets/Samsung.Tizen.Sdk.Packaging.targets#L137-L153

**Comment by espenrl** [comment]
Maybe the target names could be something along the lines of `NBGV_SetVersionForMauiAndroid`.

I'll test the functionality once there's a NuGet package on the CI feed.

**Comment by AArnott** [comment]
Sounds good. Let me know when/if you have something to work for Tizen.

**Pull request merged**